src/hotspot/share/gc/parallel/psCardTable.cpp

```diff
@@ -123,70 +124,137 @@ static void prefetch_write(void *p) {
   }
 }
 
-// postcondition: ret is a dirty card or end_card
-CardTable::CardValue* PSCardTable::find_first_dirty_card(CardValue* const start_card,
-                                                         CardValue* const end_card) {
-  for (CardValue* i_card = start_card; i_card < end_card; ++i_card) {
-    if (*i_card != PSCardTable::clean_card_val()) {
-      return i_card;
-    }
-  }
-  return end_card;
+void PSCardTable::scan_obj_with_limit(PSPromotionManager* pm,
+                                      oop obj,
+                                      HeapWord* start,
+                                      HeapWord* end) {
+  prefetch_write(start);
+  pm->push_contents_bounded(obj, start, end);
 }
 
-// postcondition: ret is a clean card or end_card
-// Note: if a part of an object is on a dirty card, all cards this object
-// resides on are considered dirty.
-CardTable::CardValue* PSCardTable::find_first_clean_card(ObjectStartArray* const start_array,
-                                                         CardValue* const start_card,
-                                                         CardValue* const end_card) {
-  assert(start_card == end_card ||
-         *start_card != PSCardTable::clean_card_val(), "precondition");
-  // Skip the first dirty card.
-  CardValue* i_card = start_card + 1;
-  while (i_card < end_card) {
-    if (*i_card != PSCardTable::clean_card_val()) {
-      i_card++;
-      continue;
+void PSCardTable::pre_scavenge(HeapWord* old_gen_bottom, uint active_workers) {
+  _pre_scavenge_active_workers = active_workers;
+}
+
+// Scavenge objects on dirty cards of the given stripe [start, end). Accesses to
+// the card table and scavenging is strictly limited to the stripe. The work on
+// objects covering multiple stripes is shared among the worker threads owning the
+// stripes.  To support this the card table is preprocessed before
+// scavenge. Imprecise dirty marks of non-objArrays are copied from start stripes
+// to all stripes (if any) they extend to.
+// A copy of card table entries corresponding to the stripe called "shadow" table
+// is used to separate card reading, clearing and redirtying.
+template <typename Func>
+void PSCardTable::process_range(Func&& object_start,
+                                PSPromotionManager* pm,
+                                HeapWord* const start,
+                                HeapWord* const end) {
+  assert(start < end, "precondition");
+  assert(is_card_aligned(start), "precondition");
+
+  // Helper struct to keep the following code compact.
+  struct Obj {
```

Comment:
The scan-dirty-chunk code below is already quite compact; this struct-def is bulky in contrast. Maybe using plain variables inside while-loop is sufficient. Not sure either is definitely superior though.

Reply:
Ok. Done.